media-proxy/src/proxy_context.cc

```diff
@@ -538,32 +537,34 @@ int ProxyContext::RxStart_rdma(const mcm_conn_param *request)
             return -EINVAL;
         }
     }
-    dp_ctx = new (dp_session_context_t);
+
     ParseMemIFParam(request, memif_ops);
     opts.transfer_size = request->payload_args.rdma_args.transfer_size;
     opts.dir = direction::RX;
     memcpy(&opts.remote_addr, &request->remote_addr, sizeof(request->remote_addr));
     memcpy(&opts.local_addr, &request->local_addr, sizeof(request->local_addr));
-    rx_ctx = rdma_rx_session_create(mDevHandle_rdma, &opts, &memif_ops);
-    if (!rx_ctx) {
-        INFO("%s, Failed to create RDMA session.", __func__);
-        delete dp_ctx;
-        return -EINVAL;
+
+    rx_rdma_session *session_ptr = NULL;
+    try {
+        session_ptr = new rx_rdma_session(mDevHandle_rdma, &opts, &memif_ops);
+    } catch (const exception &e) {
+        ERROR("%s, Failed to create RX RDMA session.", __func__);
+        ERROR("%s", e.what());
+        return -1;
     }
-    dp_ctx->payload_type = request->payload_type;
-    dp_ctx->rx_rdma_session = rx_ctx;
-    dp_ctx->id = memif_ops.m_session_count;
-    INFO("%s, session id: %d", __func__, dp_ctx->id);
-    dp_ctx->type = RX;
-    mDpCtx.push_back(dp_ctx);
 
-    return dp_ctx->id;
+    session_ptr->payload_type = request->payload_type;
+    session_ptr->id = memif_ops.m_session_count;
+    INFO("%s, session id: %d", __func__, session_ptr->id);
+    session_ptr->type = RX;
+    mDpCtx.push_back(session_ptr);
+
+    return session_ptr->id;
 }
 
 int ProxyContext::RxStart_mtl(const mcm_conn_param *request)
```

Comment:
The "case switch part" could be even a separate class as it is basically a factory pattern. Calling it explicitly a factory could help readability. 
(if we want to be very crazy, we could even use abstract factory with concrete factories for RX and TX. I guess this could also work, but this is a lot of abstraction)

Reply:
Any bigger changes to proxy_context.cc and api_server_tcp.cc should be done in the next PR to not to touch too many topics at once

Reply:
> * Break this function into a few smaller ones. for example move device init into first function, move the case switch to second, call the two of them and set variables in this one.
> * change imtl_init_preparing to some real lock
> * the same applies to TxStart_mtl
> * please add a lock similar to imtl_init_preparing (but actually working) also to RxStart_rdma and TxStart_rdma

I think the best way is to move MTL/RDMA device creation before we accept any connection(both of them), and close those devices after we stop accepting connection. This way we avoid unnecessary locks and imtl_init_preparing will disappear. Thanks to that only mDpCtx have to be locked

Reply:
I think that we can define some arguments to media_proxy to specify what backend will be initialized at startup. maybe everything availably should be initialized by default.